Add support for including vc runtime dlls

timminator opened the pull request:
# What does this PR do?

This PR allows Nuitka to also include dlls from VC++ runtime redistributables in the final standalone folder that are not guaranteed to be installed on every Windows System.

I am interested in your feedback.

# Why was it initiated? Any relevant Issues?

Fixes #3583 

# PR Checklist

- [x] Correct base branch selected? Should be `develop` branch.
- [x] Enabled commit hook or executed `./bin/autoformat-nuitka-source`.
- [x] All tests still pass. Check the Developer Manual about `Running the Tests`. There are GitHub
  Actions tests that cover the most important things however, and you are welcome to rely on those,
  but they might not cover enough.
- [x] Ideally new features or fixed regressions ought to be covered via new tests.
- [ ] Ideally new or changed features have documentation updates.

## Summary by Sourcery

Include missing Visual C++ runtime redistributable DLLs in standalone builds by discovering the MSVC Redist installation via vswhere and searching for known redistributable DLLs.

New Features:
- Discover the VC++ Redistributable installation path on Windows using vswhere
- Automatically search for and include known VCRedist DLLs (e.g., vcruntime140.dll) in the standalone output
- Retain redistributable DLLs in standalone builds even if they lie outside the Python search path

Enhancements:
- Add a helper function to recursively search for DLLs in the redistributable directory
- Emit a warning when a required VCRedist DLL cannot be found due to missing Visual Studio installation

timminator commented on nuitka/freezer/DependsExe.py:
```diff
@@ -59,26 +59,40 @@ def getDependsExePath():
-        return dll_filename
+    if is_vcredist_dll and VCREDIST_PATH:
+        # Lets attempt to find it in the MSVC Redistributable folder
+        return _find_dll_in_redist(VCREDIST_PATH, dll_filename)
```
It's just the basename already...

timminator commented on nuitka/utils/Utils.py:
```diff
@@ -508,6 +509,81 @@ def isCoffUsingPlatform():
+        if not version_folders:
+            return None
+
+        latest_version = sorted(
+            version_folders, key=lambda v: list(map(int, v.split(".")))
```
As far as im aware, this folder only contains folders with the Redistributables version name or folders with the name of the MSVC compiler version (e.g. v143) which is checked already above these lines of code.

timminator commented on nuitka/utils/Utils.py:
```diff
@@ -508,6 +509,81 @@ def isCoffUsingPlatform():
+            "*",
+            "-prerelease",
+        ]
+        vs_path = subprocess.check_output(command).decode("utf8").strip()
```
The command is not built on user input it is hardcoded right above it. We are also not using shell=True.

timminator commented on nuitka/utils/Utils.py:
```diff
@@ -508,6 +509,81 @@ def isCoffUsingPlatform():
+            "*",
+            "-prerelease",
+        ]
+        vs_path = subprocess.check_output(command).decode("utf8").strip()
```
The command is not built on user input it is hardcoded right above it. We are also not using shell=True.

kayhayen commented on nuitka/freezer/DependsExe.py:
```diff
@@ -118,14 +132,44 @@ def _parseDependsExeOutput2(lines):
-            if dll_filename.startswith("python") and dll_filename.endswith(".dll"):
-                dll_filename = _attemptToFindNotFoundDLL(dll_filename)
+            is_vcredist_dll = os.path.basename(dll_filename) in VCREDIST_DLLS
+            is_python_dll = dll_filename.startswith("python") and dll_filename.endswith(
```
Just noticing that "python" is not treated the same, i.e. no basename taken. Apparently as a DLL it's usually found either way.

timminator commented on nuitka/utils/Utils.py:
```diff
@@ -508,6 +509,81 @@ def isCoffUsingPlatform():
 
 
+def findVCRedistPath():
+    vswhere_path = os.path.join(
```
The vswhere.exe is always in the Program Files (x86) directory because the installer is only 32bit. So we do not need to check both. Otherwise the scons code seems similar to what I have done.
I understand your concern that vswhere could not be available, but I'm not sure what else we could do. We could check the hardcoded paths where VS Studio is normally installed but that could get messy quickly (e.g. several system drives etc.)...

timminator replied:
I also took a look at the scons vc.py file. It is quite a big one. From what I see for VS 2017+ they also only use the vswhere way of detecting MSVC. From what I found this also always shipped - it is also present if just the build tools were installed for example.
For older possible installations it falls back to checking the registry. For VS 2017+ checking the registry does no longer work as far as I can see as Microsoft is no longer writing install locations inside the registry.
On both my test system Visual Studio is also not adding any entries to the %PATH% variable so this will also not help us.
I think supporting VS 2017+ should be enough? VSWhere is then our best option.

timminator commented on nuitka/freezer/Standalone.py:
```diff
@@ -50,6 +50,7 @@
 from nuitka.utils.Signing import addMacOSCodeSignature
 from nuitka.utils.Timing import TimerReport
 from nuitka.utils.Utils import (
+    VCREDIST_PATH,
```
Fixed this in my new commit.

timminator commented on nuitka/freezer/DependsExe.py:
```diff
@@ -118,14 +132,44 @@ def _parseDependsExeOutput2(lines):
+                    if is_vcredist_dll:
+                        inclusion_logger.warning(
+                            """\
+                            Required Visual C++ Redistributable DLL '%s' could not be included. \
```
Fixed this in my new commit. The warning is now printed once in the end.

kayhayen commented:
Yes, I think it's a scan dir or a set of. We also wouldn't have to care about arch, and can let it handle that.

Am I getting you right there, that the vc redist lives on `%PATH%` or inside of there? I guess, I need to actually run your code or the vswhere command. Nuitka has had issues with unwanted DLLs from Tencent software in %PATH%, and as such I added restrictions to not scan outside of Python at all, unless system DLLs are allowed for. Maybe then your path just needs to be exception for that.

I may not find the time to look into this further today as I got a sponsor issue at hand, but it seems to me, that got the building blocks for a solution at least.

kayhayen commented:
So, it seems the `msvcp140_atomic_wait.dll` is only in my `C:}Windows}System32` folder, and that's also in my `%PATH%` and that's it. I did find it also from a bunch other folders. So maybe all we need to do is to declare that folder a scan_dir overriding use_path.

```
def _getPathContribution(use_path):
    """Contributions from PATH environment variable.

    These are ignored if use_path is False, but for
    Anaconda we need to keep those elements pointing
    to inside of it.
    """
```

There is an exception for Anaconda paths, and one more can be made for Windows and that system folder (still need to look it up, Nuitka does that somewhere already). 

And when then the result is produced (by any tool), so after this we just need the `_reduceToPythonPath` to allow it on Win32 for the filenames you know, and we would be good?

This of course all hinges on the fact that a VC redist installer has no other installation location than System32 folder.

And one thing not to forget, for the caching of the DLL dependencies we better make sure to consider the option you added.

kayhayen commented:
I just checked, there is no such option yet, but I think `--windows-include-redist-dlls` could be such an option. I am still pretty sure it's not legal to distribute them. But a lot of people do that, LLVM, Firefox, Android Studio, all big people, and I am not a lawyer.

timminator commented:
I'm not sure if I can follow you correctly, but we are not allowed to copy them out of the System32 that is not allowed. So we cannot add that as a scan dir. That's why I added the Microsoft Visual Studio detection and added these paths to the scan_dirs. As stated in one of the posts in the mentioned issue we are allowed to use them from there:
https://learn.microsoft.com/en-us/visualstudio/releases/2022/redistribution#visual-c-runtime-files

> Subject to the License Terms for the software, you may copy and distribute with your program any of the files within the following folder and its subfolders except as noted below. You may not modify these files.

> [VisualStudioFolder]\VC\redist

So what you want to do is way easier but definitely not allowed. So I came up with the solution implemented in this PR.

kayhayen commented:
I wonder if the Win32 SDK directory wouldn't be available by a call similar to this: ```getSconsReportValue(source_dir=source_dir, key="CCACHE_LOGFILE")```` which is how we lookup values from the scons environment at C compile time.

kayhayen commented:
But I wonder, if we can just work from this:

```
VCINSTALLDIR ['C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\VC\\']
```

When MSVC is detected, that value would be available. Right now there is this code in the 4.x inline copy of Scons, that we could change:

```
    for k, v in d.items():
        # Nuitka: Make the Windows SDK version visible in environment.
        if k == "WindowsSDKVersion":
            # Always just a single version if any.
            if len(v) == 1:
                env["WindowsSDKVersion"] = v[0].rstrip('\\')
            elif len(v) == 0:
                env["WindowsSDKVersion"] = None
            else:
                assert False, v

            continue

        print(k,v) # Added that to see what we could publish too
        assert "SDK" not in k, (k,v)

        env.PrependENVPath(k, v, delete_existing=True)
        debug("env['ENV']['%s'] = %s" % (k, env['ENV'][k]))
```

kayhayen commented:
The scons inline copy has code dealing with it:

```
class WindowsSDK(SDKDefinition):
    """
    A subclass for trying to find installed Windows SDK directories.
    """
    HKEY_FMT = r'Software\Microsoft\Microsoft SDKs\Windows\v%s\InstallationFolder'
    def __init__(self, *args, **kw):
        SDKDefinition.__init__(self, *args, **kw)
        self.hkey_data = self.version
```

But it comes down to mapping arch to folders which your code already does, and then it's good. We just need to use registry methods as per `getInstalledPythonRegistryPaths` which makes sure to use stdlib rather than pywin32. It seems easy to do this.

kayhayen commented:
Seems I am coming around to merging this as it is now, it seems reasonable to do it like this.

I wonder if for older Python, we would have to use older MSVC installations potentially, but I am choosing to not consider that to matter much now, mostly we need not bother with this. I also wonder about the Windows 7 impact.

kayhayen commented on nuitka/utils/Utils.py:
```diff
@@ -508,6 +509,91 @@ def isCoffUsingPlatform():
+    if not os.path.exists(redist_base_path):
+        return None
+
+    try:
```
That was already the function I was looking for

kayhayen commented on nuitka/freezer/Standalone.py:
```diff
@@ -51,6 +51,7 @@
 from nuitka.utils.Timing import TimerReport
 from nuitka.utils.Utils import (
     getOS,
+    getVCRedistPath,
```
I renamed a bunch of things accordingly

kayhayen commented:
But other than that, this is ready for merge now.
